lib/srv/db/mongodb/engine.go

```diff
@@ -269,7 +273,35 @@ func (e *Engine) checkClientMessage(sessionCtx *common.Session, message protocol
 			database)...)
 }
 
+func (e *Engine) waitForAnyClientMessage(clientConn net.Conn) protocol.Message {
+	clientMessage, err := protocol.ReadMessage(clientConn, e.maxMessageSize)
+	if err != nil {
+		e.Log.Warnf("Failed to read a message for reply: %v.", err)
+	}
+	return clientMessage
+}
+
 func (e *Engine) replyError(clientConn net.Conn, replyTo protocol.Message, err error) {
+	// If an error happens during server connection, wait for a client message
+	// before replying to ensure the client can interpret the reply.
+	// The first message is usually the isMaster hello message.
+	if replyTo == nil && !e.serverConnected.Load() {
+		waitChan := make(chan protocol.Message, 1)
+		go func() {
+			waitChan <- e.waitForAnyClientMessage(clientConn)
```

Comment:
If I correctly understand what's happening here, this "side event loop" will compete with the main loop inside `HandleConnection`. Even though you check the `serverConnected` field, there still is a scenario in which first, the check is performed, then `serverConnected` is set to `true`, and this line races with `HandleConnection`. Since the `protocol.ReadMessage` function is not thread-safe, this may result in an undefined behavior.

I know that this scenario may be ruled out by an assumption that `SendError` will not be called concurrently with `HandleConnection`, but then, using an atomic bool sends a different message.

I see two solutions. Perhaps it would be better to implement this part inside `HandleConnection` itself: read messages through a goroutine and a channel, `select` over that channel and another one that injects error messages and/or timeouts. And instead of waiting for a client message here, just inject an expected error to the queue and let `HandleConnection` deal with it.

Another solution is to ignore the problem (if `SendError` is not supposed to be called concurrently), but then, let's not even use the atomic boolean flag, since it just confuses the reader and conveys a wrong message about type safety assumptions; we could also explicitly state that the `Engine` methods should not expected to be type-safe.

Reply:
It's safe to assume that `SendError` is only called when `HandleConnection` fails. I've added a comment to `replyError`. As for `serverConnected`, I used atomic as I thought I would need it for the UT. Turned out it was not needed so downgraded to a regular bool.

Thanks.